File: src/BinaryTreePostorderTraversal.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;


public class BinaryTreePostorderTraversal {

    public static void main(String[] args) {
        TreeNode treeNode = TreeGenerator.generateTreeBalanced();

        BinaryTreePostorderTraversal postorderTraversal = new BinaryTreePostorderTraversal();

        Printer.printIntegerList(postorderTraversal.postorderTraversalIterative(treeNode));
    }

    public List<Integer> postorderTraversalIterative(TreeNode root) {
        TreeNode cur = root;
        Stack<TreeNode> stack = new Stack();
        List<Integer> result = new ArrayList<>();

        while (cur != null || !stack.isEmpty()) {
            if (cur != null) {
                stack.push(cur);

                // handle all left treenode.
                cur = cur.left;
            }  else {
                // check right treenode.
                TreeNode temp = stack.peek().right;
                if (temp == null) {
                    // handle last single nodes.
                    temp = stack.pop();
                    result.add(temp.val);

                    while (!stack.isEmpty() && stack.peek().right == temp) {
                        temp = stack.pop();
                        result.add(temp.val);
                    }
                }
                // has more nodes need to process.
                else {
                    cur = temp;
                }
            }
        }
        return result;
    }
}

File: data-structure/Tree/Tree/PostOrder/binary_tree_post_order_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
  """
  @param root: A Tree
  @return: Postorder in ArrayList which contains node values.
  """

  def postorderTraversal(self, root):
    # write your code here
    result = []
    stack = []

    if root is None:
      return result

    cur = root

    while cur is not None or stack:
      # handle left branch
      while cur is not None:
        stack.append(cur)
        cur = cur.left

      temp = stack[-1].right

      # handle right branch
      if temp is None:
        temp = stack.pop()
        result.append(temp.val)

        while stack and stack[-1].right == temp:
          temp = stack.pop()
          result.append(temp.val)

      else:
        cur = temp

    return result
